internal/elasticsearch/security/api_key.go

```diff
@@ -0,0 +1,241 @@
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			ValidateFunc: validation.All(
+				validation.StringLenBetween(1, 1024),
+				validation.StringMatch(regexp.MustCompile(`^([[:graph:]]| )+$`), "must contain alphanumeric characters (a-z, A-Z, 0-9), spaces, punctuation, and printable symbols in the Basic Latin (ASCII) block. Leading or trailing whitespace is not allowed"),
+			),
+		},
+		"role_descriptors": {
+			Description:      "Role descriptors for this API key.",
+			Type:             schema.TypeString,
+			Optional:         true,
+			ForceNew:         true,
+			ValidateFunc:     validation.StringIsJSON,
+			DiffSuppressFunc: utils.DiffJsonSuppress,
+		},
+		"expiration": {
+			Description: "Expiration time for the API key. By default, API keys never expire.",
+			Type:        schema.TypeString,
+			Optional:    true,
+			ForceNew:    true,
+		},
+		"expiration_timestamp": {
+			Description: "Expiration time in milliseconds for the API key. By default, API keys never expire.",
+			Type:        schema.TypeInt,
+			Computed:    true,
+		},
+		"metadata": {
+			Description:      "Arbitrary metadata that you want to associate with the API key.",
+			Type:             schema.TypeString,
+			Optional:         true,
+			Computed:         true,
+			ForceNew:         true,
+			ValidateFunc:     validation.StringIsJSON,
+			DiffSuppressFunc: utils.DiffJsonSuppress,
+		},
+		"api_key": {
+			Description: "Generated API Key.",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"encoded": {
+			Description: "API key credentials which is the Base64-encoding of the UTF-8 representation of the id and api_key joined by a colon (:).",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+	}
+
+	utils.AddConnectionSchema(apikeySchema)
+
+	return &schema.Resource{
+		Description: "Creates an API key for access without requiring basic authentication. See, https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-api-key.html",
+
+		CreateContext: resourceSecurityApiKeyCreate,
+		UpdateContext: resourceSecurityApiKeyUpdate,
+		ReadContext:   resourceSecurityApiKeyRead,
+		DeleteContext: resourceSecurityApiKeyDelete,
+
+		Schema: apikeySchema,
+	}
+}
+
+func resourceSecurityApiKeyCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client, err := clients.NewApiClient(d, meta)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	nameId := d.Get("name").(string)
+
+	var apikey models.ApiKey
+	apikey.Name = nameId
+
+	if v, ok := d.GetOk("expiration"); ok {
+		apikey.Expiration = v.(string)
+	}
+
+	if v, ok := d.GetOk("role_descriptors"); ok {
+		role_descriptors := map[string]models.Role{}
+		api_role_descriptors := map[string]models.ApiKeyRole{}
+
+		if err := json.NewDecoder(strings.NewReader(v.(string))).Decode(&role_descriptors); err != nil {
+			return diag.FromErr(err)
+		}
+
+		for key, value := range role_descriptors {
```

Comment:
Can we add a comment here about why we need this nonsense.

Reply:
Verified it works when removing the index field ⚡